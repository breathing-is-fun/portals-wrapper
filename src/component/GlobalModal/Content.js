import React, { Component } from 'react';

import Loader from '../Grid/ModulesLoader';

export default class Content extends Component {
  constructor(props) {
    super(props);

    this.state = {};

    this.root = React.createRef();
  }

  componentDidMount = () => {
    const { path } = this.props;

    if (path) {
      const loader = new Loader(
        [
          {
            i: 'content',
            path,
          },
        ],
        { content: this.root.current },
      );

      loader.load(SCTool);
    }
  };

  render = () => {
    const { content, path } = this.props;
    let renderHtml = content;

    if (path) {
      renderHtml = null;
    }

    if (!path && typeof content == 'string') {
      renderHtml = (
        <div
          dangerouslySetInnerHTML={{ __html: content }}
          style={{ width: '100%', height: '100%' }}
        />
      );
    }

    return (
      <div className="Content">
        {renderHtml}
        <div ref={this.root} />
      </div>
    );
  };
}
