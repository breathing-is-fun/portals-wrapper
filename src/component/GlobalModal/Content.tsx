import React, { Component } from 'react';

import Loader from '../Grid/ModulesLoader';

export interface ContentProps {
  path: string;
  content: any;
}
export default class Content extends Component<ContentProps> {
  private root: any;

  constructor(props: any) {
    super(props);

    this.state = {};

    this.root = React.createRef();
  }

  componentDidMount = () => {
    const { path } = this.props;

    if (path) {
      const loader = new Loader([{ i: 'globalModalContent', path }], {
        globalModalContent: this.root.current,
      });

      loader.load(SCTool);
    }
  };

  render = () => {
    const { content, path } = this.props;
    let renderHtml = content;

    if (path) {
      renderHtml = null;
    }

    if (!path && typeof content == 'string') {
      renderHtml = (
        <div
          dangerouslySetInnerHTML={{ __html: content }}
          style={{ width: '100%', height: '100%' }}
        />
      );
    }

    return (
      <div className="Content">
        {renderHtml}
        <div ref={this.root} />
      </div>
    );
  };
}
