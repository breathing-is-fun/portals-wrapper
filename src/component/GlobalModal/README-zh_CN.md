[English](./README.md) | 简体中文

# 简介

- 就像名字里写的一样，这是一个全局弹框

- 只需改变全局变量 `SCTool.modal` 的属性就可以控制弹框

- 基于 [antd的Modal](https://ant.design/components/modal-cn/)
  
# API  

| 名称 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| on | 监听的对象，会寻找对象中 `modal` 对象并监听 [可用属性](https://ant.design/components/modal-cn/#API) | object | - |

关闭时默认会销毁 `content` 中的元素

# 示例

如果在被引入插件中需要弹框，就可以这么写

``` jsx
    class Demo {
        init = () => {
            window.SCTool.modal = {
                visible: true,
                content: '<div>test</div>',
                // path: '../../../thirdModules/dbzx/index.js', // 当二级页面需要动态加载时，可以试试这个
            }
        }
    }
```