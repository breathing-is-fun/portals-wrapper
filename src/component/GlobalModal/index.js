import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Content from './Content';
import { Modal } from 'antd';

export default class GlobalModal extends Component {
  constructor (props) {
    super(props);

    this.state = {
      title: '',
      visible: false,
      content: null,
      style: {},
    };
  }

  static defaultProps = {
    on: {},
    type: 'modal',
  }

  componentDidMount = () => {
    const { on, type } = this.props;

    Object.defineProperty(on, type, {
      enumerable: true,
      configurable: true,
      set: value => {
        this.setState({ ...value });

        return value;
      }
    });
  }

  render = () => {
    const { visible, title, style, ...restStates } = this.state;

    return (
      <div className='GlobalModal'>
        <Modal
          title={title}
          visible={visible}
          style={style}
          destroyOnClose
          onCancel={() => this.setState({ visible: false })}
          footer={null}
        >
          <Content {...restStates} />
        </Modal>
      </div>
    );
  }
}

GlobalModal.propTypes = {
  on: PropTypes.object,
  type: PropTypes.string,
};