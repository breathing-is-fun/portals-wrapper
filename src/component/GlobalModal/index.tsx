import React, { Component } from 'react';
import Content from './Content';
import omit from 'omit.js';
import { Modal } from 'antd';

export interface GlobalModalProps {
  on: object;
  type?: any | 'modal';
}

export interface GlobalModalState {
  title: string;
  visible: boolean;
  content: any;
  style: object;
  mask: boolean;
  width: number | string;
  path: any;
}

export default class GlobalModal extends Component<
  GlobalModalProps,
  GlobalModalState
> {
  constructor(props: any) {
    super(props);

    this.state = {
      title: '',
      visible: false,
      content: null,
      style: {},
      mask: false,
      width: 0,
      path: '',
    };
  }

  componentDidMount = () => {
    const { on, type } = this.props;
    const defaultProps = {
      mask: false,
      width: '100%',
      style: { top: 65 },
    };

    Object.defineProperty(on, type, {
      enumerable: true,
      configurable: true,
      set: (value: any) => {
        if (!value.visible) {
          return value;
        }

        this.setState({ ...defaultProps, ...value });

        return value;
      },
    });
  };

  handleOnCancel = () => {
    this.setState({
      visible: false,
      path: null,
      content: null,
    });
  };

  render = () => {
    const { visible, title, style, width, ...restStates } = this.state;
    const newStates = omit(this.state, ['content', 'path', 'style']);

    return (
      <div className="GlobalModal">
        <Modal
          destroyOnClose
          onCancel={this.handleOnCancel}
          footer={null}
          forceRender
          style={Object.assign({}, { top: 65 }, style)}
          {...newStates}
        >
          <Content {...restStates} />
        </Modal>
      </div>
    );
  };
}
