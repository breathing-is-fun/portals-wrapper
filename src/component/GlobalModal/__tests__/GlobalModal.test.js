import React from 'react';
import { render, mount, shallow } from 'enzyme';
import GlobalModal from '..';

describe('GlobalModal', () => {
  let tool = {};
  const demo = <GlobalModal on={ tool } />;
  const wrapper = mount(demo);

  it('when visible is true, Modal should work', () => {
    tool.modal = {
      visible: true,
    };

    expect(wrapper).toMatchSnapshot();
  });

  it('when content exists, dangerouslySetInnerHTML should work', () => {
    tool.modal = {
      visible: true,
      content: '<div>test</div>',
    };

    expect(wrapper).toMatchSnapshot();

    tool.modal = {
      content: <div>test</div>,
    };

    expect(wrapper).toMatchSnapshot();
  });
});