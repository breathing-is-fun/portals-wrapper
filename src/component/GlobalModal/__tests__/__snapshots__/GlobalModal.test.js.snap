// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalModal when content exists, dangerouslySetInnerHTML should work 1`] = `
<GlobalModal
  on={
    Object {
      "modal": undefined,
    }
  }
  type="modal"
>
  <div
    className="GlobalModal"
  >
    <Modal
      cancelButtonDisabled={false}
      confirmLoading={false}
      destroyOnClose={true}
      footer={null}
      maskTransitionName="fade"
      okButtonDisabled={false}
      okType="primary"
      onCancel={[Function]}
      prefixCls="ant-modal"
      style={Object {}}
      title=""
      transitionName="zoom"
      visible={false}
      width={520}
    >
      <DialogWrap
        cancelButtonDisabled={false}
        closeIcon={
          <span
            className="ant-modal-close-x"
          >
            <Icon
              className="ant-modal-close-icon"
              type="close"
            />
          </span>
        }
        confirmLoading={false}
        destroyOnClose={true}
        footer={null}
        maskTransitionName="fade"
        okButtonDisabled={false}
        okType="primary"
        onCancel={[Function]}
        onClose={[Function]}
        prefixCls="ant-modal"
        style={Object {}}
        title=""
        transitionName="zoom"
        visible={false}
        width={520}
        wrapClassName=""
      />
    </Modal>
  </div>
</GlobalModal>
`;

exports[`GlobalModal when content exists, dangerouslySetInnerHTML should work 2`] = `
<GlobalModal
  on={
    Object {
      "modal": undefined,
    }
  }
  type="modal"
>
  <div
    className="GlobalModal"
  >
    <Modal
      cancelButtonDisabled={false}
      confirmLoading={false}
      destroyOnClose={true}
      footer={null}
      maskTransitionName="fade"
      okButtonDisabled={false}
      okType="primary"
      onCancel={[Function]}
      prefixCls="ant-modal"
      style={Object {}}
      title=""
      transitionName="zoom"
      visible={false}
      width={520}
    >
      <DialogWrap
        cancelButtonDisabled={false}
        closeIcon={
          <span
            className="ant-modal-close-x"
          >
            <Icon
              className="ant-modal-close-icon"
              type="close"
            />
          </span>
        }
        confirmLoading={false}
        destroyOnClose={true}
        footer={null}
        maskTransitionName="fade"
        okButtonDisabled={false}
        okType="primary"
        onCancel={[Function]}
        onClose={[Function]}
        prefixCls="ant-modal"
        style={Object {}}
        title=""
        transitionName="zoom"
        visible={false}
        width={520}
        wrapClassName=""
      />
    </Modal>
  </div>
</GlobalModal>
`;

exports[`GlobalModal when visible is true, Modal should work 1`] = `
<GlobalModal
  on={
    Object {
      "modal": undefined,
    }
  }
  type="modal"
>
  <div
    className="GlobalModal"
  >
    <Modal
      cancelButtonDisabled={false}
      confirmLoading={false}
      destroyOnClose={true}
      footer={null}
      maskTransitionName="fade"
      okButtonDisabled={false}
      okType="primary"
      onCancel={[Function]}
      prefixCls="ant-modal"
      style={Object {}}
      title=""
      transitionName="zoom"
      visible={false}
      width={520}
    >
      <DialogWrap
        cancelButtonDisabled={false}
        closeIcon={
          <span
            className="ant-modal-close-x"
          >
            <Icon
              className="ant-modal-close-icon"
              type="close"
            />
          </span>
        }
        confirmLoading={false}
        destroyOnClose={true}
        footer={null}
        maskTransitionName="fade"
        okButtonDisabled={false}
        okType="primary"
        onCancel={[Function]}
        onClose={[Function]}
        prefixCls="ant-modal"
        style={Object {}}
        title=""
        transitionName="zoom"
        visible={false}
        width={520}
        wrapClassName=""
      />
    </Modal>
  </div>
</GlobalModal>
`;
